### PR TITLE
Profiles in treeview

### PR DESCRIPTION
**Pull request by printpagestopdf**
Shows Profiles in treeview with all editing capabilities.
Supports copy/cut/paste by keyboard and mouse. Inline editing for renaming and new.

Ftp tree although supports copy/cut/paste (without copying directories) and inline renaming.

For a preview the 64bit dll
https://github.com/printpagestopdf/NppFTP/releases/download/v0.1-beta/NppFTP.dll

### CODE REVIEW

**Comment by printpagestopdf** [comment]
@ashkulz is there something I have to change or what is the next step in the procedure?

**Comment by chcg** [comment]
A build is also available here https://ci.appveyor.com/project/ashkulz/nppftp/builds/34182354/job/5klgxkvc0yycrd5y/artifacts

File: src/FTPProfile.cpp [hunk 1]
```diff
@@ -51,7 +51,6 @@ FTPProfile::FTPProfile() :
 }
 
 FTPProfile::FTPProfile(const TCHAR * name) :
-	m_parent(NULL),
 	m_port(21),
```
Comment (chcg) [inline src/FTPProfile.cpp:54]: why is this necessary?

File: src/FTPProfile.cpp [hunk 2]
```diff
@@ -67,7 +66,7 @@ FTPProfile::FTPProfile(const TCHAR * name) :
 	m_cache = new FTPCache();
 
 	m_name = SU::DupString(name);
-	m_parent = SU::DupString(L"");
+	m_parent = SU::DupString(TEXT(""));
```
Comment (chcg) [inline src/FTPProfile.cpp:69]: why is this necessary?

File: src/FTPProfile.cpp [hunk 3]
```diff
@@ -82,7 +81,6 @@ FTPProfile::FTPProfile(const TCHAR * name) :
 }
 
 FTPProfile::FTPProfile(const TCHAR * name, const FTPProfile* other) :
-	m_parent(other->m_parent),
 	m_port(other->m_port),
```
Comment (chcg) [inline src/FTPProfile.cpp:84]: why is this necessary?

File: src/FTPProfile.cpp [hunk 4]
```diff
@@ -102,6 +100,7 @@ FTPProfile::FTPProfile(const TCHAR * name, const FTPProfile* other) :
 	m_username = SU::strdup(other->m_username);
 	m_password = SU::strdup(other->m_password);
 	m_initialDir = SU::strdup(other->m_initialDir);
+	m_parent = SU::DupString(other->m_parent);
```
Comment (chcg) [inline src/FTPProfile.cpp:103]: why is this necessary?

File: src/FileObject.cpp [hunk 5]
```diff
@@ -135,7 +137,7 @@ int FileObject::SetParent(FileObject * parent) {
 }
 
 FileObject* FileObject::GetParent() {
-	return m_parent;
+	return m_parent;
```
Comment (chcg) [inline src/FileObject.cpp:140]: Added superfluous whitespace

File: src/FileObject.h [hunk 6]
```diff
@@ -65,6 +67,7 @@ class FileObject {
 	virtual int				Sort();
 
 	static int				SortVector(FOVector & foVect);
+	WORD					m_cutpaste;
```
Comment (chcg) [inline src/FileObject.h:70]: Is this variable used?

File: src/QueueOperation.h [hunk 7]
```diff
@@ -142,21 +142,39 @@ class QueueDownload : public QueueOperation {
 
 class QueueDownloadHandle : public QueueOperation {
 public:
-							QueueDownloadHandle(HWND hNotify, const char * externalFile, HANDLE hFile, Transfer_Mode tMode, int notifyCode = 0, void * notifyData = NULL);
+	QueueDownloadHandle(HWND hNotify, const char* externalFile, HANDLE hFile, Transfer_Mode tMode, int notifyCode = 0, void* notifyData = NULL);
```
Comment (chcg) [inline src/QueueOperation.h:145]: In general try to avoid adding formatting stuff as part of a feature PR. So unrelated changes doesn't disturb the review of the relevant parts of the code.

File: src/QueueOperation.h [hunk 8]
```diff
@@ -46,7 +46,7 @@ class QueueOperation {
 	enum QueueType { QueueTypeConnect, QueueTypeDisconnect, QueueTypeDownload, QueueTypeUpload,
 	                 QueueTypeDirectoryGet, QueueTypeDirectoryCreate, QueueTypeDirectoryRemove,
 	                 QueueTypeFileCreate, QueueTypeFileDelete, QueueTypeFileRename, QueueTypeQuote,
-	                 QueueTypeDownloadHandle
+	                 QueueTypeDownloadHandle,QueueTypeCopyFile
```
Comment (chcg) [inline src/QueueOperation.h:49]: Add a space to be consistent with the formatting of the other types.

File: src/Windows/ProfilesDialog.cpp [hunk 9]
```diff
@@ -117,7 +137,7 @@ INT_PTR ProfilesDialog::OnCommand(int ctrlId, int notifCode, HWND idHwnd) {
 				m_ftpWindow->OnProfileChange();
 				LoadProfiles();
 			}
-			break; }
+			break; } 
 		case IDC_BUTTON_PROFILE_DELETE: {
```
Comment (chcg) [inline src/Windows/ProfilesDialog.cpp:141]: Added superfluous whitespace

File: src/Windows/ProfilesDialog.cpp [hunk 10]
```diff
@@ -592,9 +612,16 @@ INT_PTR ProfilesDialog::OnInitDialog() {
 	}
 
 	if (m_profiles->size() > 0) {
+		size_t index = 0;
+		for (size_t i = 0; i < m_profiles->size(); i++) {
```
Comment (chcg) [inline src/Windows/ProfilesDialog.cpp:616]: Would it be possible to use find instead?

File: src/FTPProfile.cpp (hunk 1 above)
Comment (printpagestopdf) [reply src/FTPProfile.cpp:54]: Its initialized in method body, but you are right, should be initialized before

File: src/FTPProfile.cpp (hunk 2 above)
Comment (printpagestopdf) [reply src/FTPProfile.cpp:69]: No good reason, changed it

File: src/FTPProfile.cpp (hunk 3 above)
Comment (printpagestopdf) [reply src/FTPProfile.cpp:84]: needed a real string copy, not only reference - readded it but with string duplication

File: src/FTPProfile.cpp (hunk 4 above)
Comment (printpagestopdf) [reply src/FTPProfile.cpp:103]: removed, added it to initialization

File: src/FileObject.cpp (hunk 5 above)
Comment (printpagestopdf) [reply src/FileObject.cpp:140]: removed

File: src/FileObject.h (hunk 6 above)
Comment (printpagestopdf) [reply src/FileObject.h:70]: Yes, if an object is part of a drag operation it holds the information if it is a cut or copy operation

File: src/QueueOperation.h (hunk 7 above)
Comment (printpagestopdf) [reply src/QueueOperation.h:145]: changed back

File: src/QueueOperation.h (hunk 8 above)
Comment (printpagestopdf) [reply src/QueueOperation.h:49]: ok

File: src/Windows/ProfilesDialog.cpp (hunk 9 above)
Comment (printpagestopdf) [reply src/Windows/ProfilesDialog.cpp:141]: ok

File: src/Windows/ProfilesDialog.cpp (hunk 10 above)
Comment (printpagestopdf) [reply src/Windows/ProfilesDialog.cpp:616]: Haven't thought about this, but seems to be the better way. Changed it accordingly.

**Comment by printpagestopdf** [comment]
Fixed errors/suggestions found by review from @chcg
